### PR TITLE
Letter-precise html tokenization

### PR DESCRIPTION
**Pull request by whalebot-helmsman**
For loseless html detokenization we should  

- store positions of our text tokens in source html file
- pass it to html tokens 
- use this information during detokenization

Also we have to divide tokenization and START/END tag cleanup process, as we want tokens from clean and tagged tree in same time

### CODE REVIEW

File: webstruct/html_tokenizer.py [hunk 1]
```diff
@@ -206,30 +215,47 @@ def detokenize_single(self, html_tokens, tags):
...
+                continue
+
+            elem = html_tokens[g[0][0]].elem
+
+            pos_in_source = 0
+            source = elem.text
+            if is_tail:
+                source = elem.tail
+
+            modded = ''
```
Comment (kmike) [inline webstruct/html_tokenizer.py]: Could you please use a list here, and join it at the end? The current code is not O(N^2) only in CPython.

File: webstruct/html_tokenizer.py [hunk 2]
```diff
@@ -41,6 +45,8 @@ class HtmlToken(_HtmlToken):
     * :attr:`elem` is the current html block (as lxml's Element) - most
       likely you want :attr:`parent` instead of it
     * :attr:`is_tail` flag indicates that token belongs to element tail
+    * :attr:`position is position of token start in parent text
+    * :attr:`length is length of token in parent text
```
Comment (kmike) [inline webstruct/html_tokenizer.py]: Let's clarify if we're talking about bytes positions or unicode positions

File: webstruct/text_tokenizers.py [hunk 3]
```diff
@@ -79,46 +143,66 @@ class WordTokenizer(object):
...
+        return [t for t in self._tokenize(text) if t.chars]
 
 
 class DefaultTokenizer(WordTokenizer):
     def tokenize(self, text):
         tokens = super(DefaultTokenizer, self).tokenize(text)
         # remove standalone commas and semicolons
-        # as they broke tag sets, e.g. PERSON->FUNCTION in case "PERSON, FUNCTION"
+        # as they broke tag sets
+        # , e.g. PERSON->FUNCTION in case "PERSON, FUNCTION"
```
Comment (kmike) [inline webstruct/text_tokenizers.py]: a nitpick: comma should be on a line above :)

File: webstruct/text_tokenizers.py [hunk 4]
```diff
@@ -79,46 +143,66 @@ class WordTokenizer(object):
...
-                        yield token
+                        yield TextToken(chars=token,
+                                        position=i + match.start(),
+                                        length=shift)
                     break
             i += shift
 
     def tokenize(self, text):
-        return [t for t in self._tokenize(text) if t]
+        return [t for t in self._tokenize(text) if t.chars]
```
Comment (kmike) [inline webstruct/text_tokenizers.py]: Does it make https://github.com/scrapinghub/webstruct/pull/36/files obsolete?

Also, it seems we can easily make this backwards compatible by using another name for tokenize, and leaving existing tokenize method as-is.

File: webstruct/sequence_encoding.py [hunk 5]
```diff
@@ -81,13 +89,14 @@ def iter_encode(self, input_tokens):
...
-        if not res:
-            return (), ()
-        tokens, tags = zip(*res)
-        return list(tokens), list(tags)
+    def split(self, tokens):
+        """ split ``[(token, tag)]`` to ``([token], [tags])`` tuple """
+        return [t[0] for t in tokens], [t[1] for t in tokens]
+
+    @classmethod
+    def from_indicies(Cls, indicies, input_tokens):
```
Comment (kmike) [inline webstruct/sequence_encoding.py]: * a typo: it should be `from_indices`;
* argument name should be `cls` according to pep8

File: webstruct/sequence_encoding.py [hunk 6]
```diff
@@ -11,23 +11,31 @@ class IobEncoder(object):
 
         >>> iob_encoder = IobEncoder()
         >>> input_tokens = ["__START_PER__", "John", "__END_PER__", "said"]
-        >>> iob_encoder.encode(input_tokens)
+        >>> [p for p in IobEncoder.from_indicies(iob_encoder.encode(input_tokens), input_tokens)]
         [('John', 'B-PER'), ('said', 'O')]
 
     Get the result in another format using ``encode_split`` method::
```
Comment (kmike) [inline webstruct/sequence_encoding.py]: `encode_split` method is removed.

File: webstruct/sequence_encoding.py [hunk 7]
```diff
@@ -11,23 +11,31 @@ class IobEncoder(object):
...
         >>> tokens, tags
         (['hello', 'John', 'Doe', 'Mary', 'said'], ['O', 'B-PER', 'I-PER', 'B-PER', 'O'])
 
     Note that IobEncoder is stateful. This means you can encode incomplete
     stream and continue the encoding later::
 
         >>> iob_encoder = IobEncoder()
-        >>> iob_encoder.encode(["__START_PER__", "John"])
+        >>> input_tokens_partial = ["__START_PER__", "John"]
+        >>> tokens = iob_encoder.encode(input_tokens_partial)
```
Comment (kmike) [inline webstruct/sequence_encoding.py]: .encode method no longer returns tokens, so it could be better to rename the variable.

File: webstruct/sequence_encoding.py [hunk 8]
```diff
@@ -36,7 +44,7 @@ class IobEncoder(object):
 
     Group results to entities::
 
-        >>> iob_encoder.group(iob_encoder.encode(input_tokens))
+        >>> iob_encoder.group([p for p in IobEncoder.from_indicies(iob_encoder.encode(input_tokens), input_tokens)])
```
Comment (kmike) [inline webstruct/sequence_encoding.py]: `IobEncoder.from_indicies(iob_encoder.encode(input_tokens), input_tokens)` is repeated in all examples. I wonder if we should keep `encode` method backwards compatible, and introduce another method, which uses indices; `encode` implementation would be `IobEncoder.from_indicies(iob_encoder.encode(input_tokens), input_tokens)`.

**Comment by kmike** [comment]
Keeping positions of text and html tokens is a great addition 👍

File: webstruct/text_tokenizers.py (hunk 4 above)
Comment (whalebot-helmsman) [reply webstruct/text_tokenizers.py]: Yes, it does same what span want to do. But 36th pull request also contains some regexp modifications.

You want to move new code to separate method, e.g. span_tokenize, and keep old method tokenize calling for span_tokenize internally, which will be return only text tokens? This is possible, but there is no consumers for old method.

Comment (kmike) [reply webstruct/text_tokenizers.py]: I think we can keep regexp modifications out of this PR; #36 was not merged because quality was decreasing a bit after tokenization changes.

As for consumers, that's true there are no consumers in webstruct, but we never know as it is open source :) Also, tokenizer is made to be compatible with nltk's tokenizer; it is not a good design to return a different data type in an overridden method.

File: webstruct/sequence_encoding.py (hunk 8 above)
Comment (whalebot-helmsman) [reply webstruct/sequence_encoding.py]: This pattern is repeated in test cases only, we can define a function in tests code.

File: webstruct/text_tokenizers.py [hunk 9]
```diff
@@ -12,9 +12,9 @@ class WordTokenizer(object):
 
     >>> from nltk.tokenize.treebank import TreebankWordTokenizer  # doctest: +SKIP
     >>> s = '''Good muffins cost $3.88\nin New York. Email: muffins@gmail.com'''
-    >>> TreebankWordTokenizer().tokenize(s) # doctest: +SKIP
+    >>> TreebankWordTokenizer().span_tokenize(s) # doctest: +SKIP
```
Comment (kmike) [inline webstruct/text_tokenizers.py]: span_tokenize returns a different kind of output, it can be a good time to either fix tests, or remove them

File: webstruct/text_tokenizers.py [hunk 10]
```diff
@@ -12,9 +12,9 @@ class WordTokenizer(object):
...
     ['Good', 'muffins', 'cost', '$', '3.88', 'in', 'New', 'York.', 'Email', ':', 'muffins', '@', 'gmail.com']
-    >>> WordTokenizer().tokenize(s)
+    >>> WordTokenizer().span_tokenize(s)
```
Comment (kmike) [inline webstruct/text_tokenizers.py]: in nltk span_tokenize returns (start, end) tuples; it can be better to either implemente a compatible API or use a different name.

File: webstruct/html_tokenizer.py [hunk 11]
```diff
@@ -266,16 +311,23 @@ def _cleanup_elem(self, elem):
             elem.tail = self._tag_re.sub("", elem.tail)
 
     def _tokenize_and_split(self, text):
-        input_tokens = self._limit_tags(self.text_tokenize_func(text or ''))
-        input_tokens = map(six.text_type, input_tokens)
-        return self.sequence_encoder.encode_split(input_tokens)
+        text = text or ''
+        input_tokens = [t for t in self.text_tokenize_func(text)]
+        input_tokens = self._limit_tags(input_tokens)
+        input_tokens = [TextToken(chars=six.text_type(t.chars),
```
Comment (kmike) [inline webstruct/html_tokenizer.py]: unicode doesn't look right here; if t.chars if unicode (str in Python 3) then conversion is not needed; if t.chars is bytes, then conversion should use a proper encoding, not sys.getdefaultencoding() which is often ascii

Comment (whalebot-helmsman) [reply webstruct/html_tokenizer.py]: This conversion is the same as it was [before](https://github.com/scrapinghub/webstruct/blob/c2af27808c27a5ffc8cdd40c921408aa9d3dcfc6/webstruct/html_tokenizer.py#L270)

Comment (whalebot-helmsman) [reply webstruct/html_tokenizer.py]: One of [tests](https://github.com/scrapinghub/webstruct/blob/c2af27808c27a5ffc8cdd40c921408aa9d3dcfc6/webstruct/tests/test_html_tokenizer.py#L59) awaits unicode

Comment (whalebot-helmsman) [reply webstruct/html_tokenizer.py]: All real encoding/decoding handled by lxml. lxml uses utf-8 as it is internal representation. I think we can add test with real unicode and safely remove this conversion

Comment (kmike) [reply webstruct/html_tokenizer.py]: Ah, I finally recalled why do we have such code! In Python 2.x lxml returns bytes for ASCII-only data and unicode for non-ascii data; this code ensures everything is unicode. It is only active for ascii-only bytes in Python 2.x, and no-op in all other cases, so it works as intended. Sorry for a false alarm.

File: webstruct/text_tokenizers.py [hunk 12]
```diff
@@ -1,61 +1,139 @@
...
-
-    >>> WordTokenizer().tokenize("Powai Campus, Mumbai-400077")  # doctest: +SKIP
-    ['Powai', 'Campus', ',', 'Mumbai", "-", "400077']
-
-    >>> WordTokenizer().tokenize("1 5858/ 1800")  # doctest: +SKIP
-    ['1', '5858', '/', '1800']
-
-    >>> WordTokenizer().tokenize("Saudi Arabia-")  # doctest: +SKIP
-    ['Saudi', 'Arabia', '-']
+    >>> WordTokenizer().segment_words("Phone:855-349-1914")
```
Comment (kmike) [inline webstruct/text_tokenizers.py]: I don't think we should enable these tests, they are like `@xfail`

**Comment by kmike** [comment]
It looks good, besides a minor comment about tests. 
Do you know how much is speed affected? Tokenization used to be one of the bottlenecks.

File: webstruct/text_tokenizers.py (hunk 12 above)
Comment (whalebot-helmsman) [reply webstruct/text_tokenizers.py]: So we should remove it at all

**Comment by whalebot-helmsman** [comment]
I don't how speed is affected. Do we have some kind of benchmark?

File: webstruct/text_tokenizers.py (hunk 12 above)
Comment (kmike) [reply webstruct/text_tokenizers.py]: I think they are nice test cases, so I'd prefer to keep them in some form, maybe convert them to pytest.mark.xfail tests.

**Comment by kmike** [comment]
There is no benchmark, but you can load all trees in our dataset and check how long does it take to tokenize them all before and after the change (see https://github.com/scrapinghub/webstruct/issues/13 or https://github.com/scrapinghub/webstruct/pull/15).

**Comment by whalebot-helmsman** [comment]
On proposed benchmark new version is 10% slower than old

```
[nikita@dsc-dev-01:~/dev/webstruct]$ runinenv.sh ~/ves/webstruct python -m webstruct.html_tokenizer_benchmark
Executing python -m webstruct.html_tokenizer_benchmark in /home/nikita/ves/webstruct
22.220800561946817

[nikita@dsc-dev-01:~/dev/webstruct]$ runinenv.sh ~/ves/webstruct python -m webstruct.html_tokenizer_benchmark
Executing python -m webstruct.html_tokenizer_benchmark in /home/nikita/ves/webstruct
24.135991828050464
```

File: webstruct/tests/test_text_tokenizer.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+class TestTokenizerTest(unittest.TestCase):
+    def do_tokenize(self, text, result):
+        self.assertEqual(result, WordTokenizer().segment_words(text))
+
+    @pytest.mark.xfail
+    def test_phone(self):
+        return self.do_tokenize(
+                "Phone:855-349-1914",
+                [TextToken(chars='Phone:855-349-1914', position=0, length=18)]
```
Comment (kmike) [inline webstruct/tests/test_text_tokenizer.py]: This is not the output we're expecting, phone should be separated (like it was in old doctests)

**Comment by kmike** [comment]
Thanks @whalebot-helmsman! I think the slowdown is tolerable, and having exact positions would make a foundation for other features, e.g. we can recover entity text in a smarter way when extracting them, or use commas as features without unconditionally removing them in tokenizer.
